_posts/2020-07-31-fnn-vae-for-noisy-timeseries/fnn_vae_for_noisy_timeseries.Rmd

```diff
@@ -0,0 +1,509 @@
+                                strides = 2)
+    self$act3 <- layer_activation_leaky_relu()
+    self$batchnorm3 <- layer_batch_normalization()
+    self$conv4 <- layer_conv_1d(
+      kernel_size = 9,
+      filters = n_latent,
+      strides = 2,
+      activation = "linear" 
+    )
+    self$batchnorm4 <- layer_batch_normalization()
+    self$flat <- layer_flatten()
+    
+    function (x, mask = NULL) {
+      x %>%
+        self$conv1() %>%
+        self$act1() %>%
+        self$batchnorm1() %>%
+        self$conv2() %>%
+        self$act2() %>%
+        self$batchnorm2() %>%
+        self$conv3() %>%
+        self$act3() %>%
+        self$batchnorm3() %>%
+        self$conv4() %>%
+        self$batchnorm4() %>%
+        self$flat()
+    }
+  })
+}
+
+```
+
+The decoder starts from this -- flat -- representation and decompresses it into a time sequence. In both encoder and decoder
+(de-) conv layers, parameters are chosen to handle a sequence length (`num_timesteps`) of 120, which is what we'll use for
+prediction below.
+
+```{r, eval = FALSE}
+vae_decoder_model <- function(n_timesteps,
+                               n_features,
+                               n_latent,
+                               name = NULL) {
+  keras_model_custom(name = name, function(self) {
+    self$reshape <- layer_reshape(target_shape = c(1, n_latent))
+    self$conv1 <- layer_conv_1d_transpose(kernel_size = 15,
+                                          filters = 64,
+                                          strides = 3)
+    self$act1 <- layer_activation_leaky_relu()
+    self$batchnorm1 <- layer_batch_normalization()
+    self$conv2 <- layer_conv_1d_transpose(kernel_size = 11,
+                                          filters = 32,
+                                          strides = 3)
+    self$act2 <- layer_activation_leaky_relu()
+    self$batchnorm2 <- layer_batch_normalization()
+    self$conv3 <- layer_conv_1d_transpose(
+      kernel_size = 9,
+      filters = 16,
+      strides = 2,
+      output_padding = 1
+    )
+    self$act3 <- layer_activation_leaky_relu()
+    self$batchnorm3 <- layer_batch_normalization()
+    self$conv4 <- layer_conv_1d_transpose(
+      kernel_size = 7,
+      filters = 1,
+      strides = 1,
+      activation = "linear"
+    )
+    self$batchnorm4 <- layer_batch_normalization()
+    
+    function (x, mask = NULL) {
+      x %>%
+        self$reshape() %>%
+        self$conv1() %>%
+        self$act1() %>%
+        self$batchnorm1() %>%
+        self$conv2() %>%
+        self$act2() %>%
+        self$batchnorm2() %>%
+        self$conv3() %>%
+        self$act3() %>%
+        self$batchnorm3() %>%
+        self$conv4() %>%
+        self$batchnorm4()
+    }
+  })
+}
+```
+
+Note that even though we called these constructors `vae_encoder_model()` and `vae_decoder_model()`, there is nothing
+variational to these models per se; they really just are an encoder and a decoder, respectively. Metamorphosis into a VAE will
```

Comment:
```suggestion
variational to these models per se; they are really just an encoder and a decoder, respectively. Metamorphosis into a VAE will
```

Reply:
I think this is one of those word order things in English where there's probably no rule, but I'm going on _sprachgefühl_.

Reply:
top :-)